File: android/hyperid_sdk/src/main/java/ai/hyper_id/sdk/hyper_id_sdk.kt
package ai.hyper_id.sdk

import ai.hyper_id.sdk.api.auth.IHyperIDSDKAuth
import ai.hyper_id.sdk.api.auth.model.ClientInfo
import ai.hyper_id.sdk.api.auth.model.HyperIdProvider
import ai.hyper_id.sdk.api.kyc.IHyperIDSDKKYC
import ai.hyper_id.sdk.api.mfa.IHyperIdSDKMFA
import ai.hyper_id.sdk.api.storage.IHyperIDSDKStorage
import ai.hyper_id.sdk.internal.HyperIdSDKImpl

fun HyperIdSDKInstance() : IHyperIdSDK = HyperIdSDKImpl()

interface IHyperIdSDK
{
	enum class RequestResult
	{
		FAIL_INIT_REQUIRED,
		FAIL_AUTHORIZATION_REQUIRED,
		FAIL_CONNECTION,
		FAIL_SERVICE,

		SUCCESS,
	}

	interface IRequestResultListener {
		fun onRequestComplete(_result		: RequestResult,
							  _errorDesc	: String?)
	}

	/**
	 * async Init SDK
	 *
	 * @param	provider			Provide full information about service connection endpoint
	 * @param	clientInfo			Provide full information about client application
	 * @param	authRestoreInfo		Optional. Could be provided to avoid authorization request
	 * @param	completeListener	Callback to notify init result
	 **/
	fun init(provider			: HyperIdProvider,
			 clientInfo			: ClientInfo,
			 authRestoreInfo	: String? = null,
			 completeListener	: IRequestResultListener)
	/**
	 * Done SDK
	 **/
	fun done()

	fun getAuth()		: IHyperIDSDKAuth
	fun getMFA()		: IHyperIdSDKMFA
	fun getKYC()		: IHyperIDSDKKYC
	fun getStorage()	: IHyperIDSDKStorage
}

File: cs/auth/hyper_id_sdk.cs
using HyperId.Private;
using HyperId.SDK.Authorization;
using HyperId.SDK.KYC;
using HyperId.SDK.MFA;
using HyperId.SDK.Storage;
using System.Diagnostics.CodeAnalysis;
using System.Threading.Tasks;
using System.Threading;

namespace HyperId.SDK
{
    public class HyperIDSDKFactory
    {
        public static IHyperIDSDK Instance() { return new HyperIDSDKImpl(); }
    }

    public interface IHyperIDSDK
    {
        /// <summary>
        /// 
        /// </summary>
        /// <param name="providerInfo">Required. provider connection params</param>
        /// <param name="clientInfo">Required. HyperId client params</param>
        /// <param name="authRestoreInfo">Optional. AuthRestoreInfo to restore SDK state</param>
        /// <returns></returns>
        /// <exception cref="TaskCanceledException"></exception>
        /// <exception cref="HyperIDAuthException"></exception>
        /// <exception cref="ArgumentNullException">Raised if <paramref name="providerInfo"/> or <paramref name="clientInfo"/> are null</exception>
        Task InitAsync(
            [NotNull] ProviderInfo providerInfo,
            [NotNull] ClientInfo clientInfo,
            [AllowNull] string? authRestoreInfo,
            CancellationToken cancellationToken = default);

        void Done();

        /// <summary>
        /// Auth Init and authrize is required to correct work with MFA/KYC/Storage modules. 
        /// </summary>
        /// <returns>interface to work with authorization SDK</returns>
        public IHyperIDSDKAuth GetAuth();
        public IHyperIDSDKMFA GetMFA();
        public IHyperIDSDKKyc GetKYC();
        public IHyperIDSDKStorage GetStorage();

        public string? GetAuthRestoreInfo();
    }

}//namespace HyperId.SDK
